pkg/counter/cache.go

```diff
@@ -0,0 +1,162 @@
+		}
+	})
+}
+
+// IncrementUserAccountCount increments the number of UserAccount fo the given member cluster in the cached counter
+func IncrementUserAccountCount(clusterName string) {
+	write(func() {
+		cachedCounts.UserAccountsPerClusterCounts[clusterName]++
+	})
+}
+
+// DecrementUserAccountCount decreases the number of UserAccount fo the given member cluster in the cached counter
+func DecrementUserAccountCount(log logr.Logger, clusterName string) {
+	write(func() {
+		if cachedCounts.UserAccountsPerClusterCounts[clusterName] != 0 {
+			cachedCounts.UserAccountsPerClusterCounts[clusterName]--
+		} else {
+			log.Error(fmt.Errorf("the count of UserAccounts is zero"),
+				"unable to decrement the number of UserAccounts for the given cluster", "cluster", clusterName)
+		}
+	})
+}
+
+// GetCounts returns Counts struct containing number of MURs and number of UserAccounts per member cluster.
+// If the counter is not yet initialized, then it returns error
+func GetCounts() (Counts, error) {
+	cachedCounts.RLock()
+	defer cachedCounts.RUnlock()
+	if !cachedCounts.initialized {
+		return cachedCounts.Counts, fmt.Errorf("counter is not initialized")
+	}
+	return cachedCounts.Counts, nil
+}
+
+// Synchronize synchronizes the content of the ToolchainStatus with the cached counter
+//
+// If the counter hasn't been initialized yet, then it adds the actual cached count
+// to the one taken from ToolchainStatus and marks the cached as initialized
+//
+// If the ToolchainStatus doesn't contain any MUR counts, then it lists all existing
+// MURs, counts them and stores in both cache and ToolchainStatus object
+//
+// If the cached counter is initialized and ToolchainStatus contains already some numbers
+// then it updates the ToolchainStatus numbers with the one taken from the cached counter
+func Synchronize(cl client.Client, toolchainStatus *v1alpha1.ToolchainStatus) error {
+	cachedCounts.Lock()
+	defer cachedCounts.Unlock()
+	if toolchainStatus.Status.HostOperator == nil || toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount == 0 {
+		if err := loadCurrentNumbers(cl, toolchainStatus.Namespace); err != nil {
+			return err
+		}
+	}
+	if toolchainStatus.Status.HostOperator == nil {
+		toolchainStatus.Status.HostOperator = &v1alpha1.HostOperatorStatus{}
+	}
+	if !cachedCounts.initialized {
+		cachedCounts.MasterUserRecordCount += toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount
+		for _, memberStatus := range toolchainStatus.Status.Members {
+			cachedCounts.UserAccountsPerClusterCounts[memberStatus.ClusterName] += memberStatus.CapacityUsage.UserAccountCount
+		}
+		cachedCounts.initialized = true
+	}
+
+	toolchainStatus.Status.HostOperator.CapacityUsage.MasterUserRecordCount = cachedCounts.MasterUserRecordCount
+CachedCountsPerCluster:
+	for clusterName, count := range cachedCounts.UserAccountsPerClusterCounts {
+		for index, member := range toolchainStatus.Status.Members {
```

Comment:
the code here could be a bit simpler if `toolchainStatus.Status.Members` was a `map[string]Member` instead of a `[]Member` 🤔

Reply:
yeah, but I don't want to refactor the `ToolchainStatus` CRD just because of this for cycle :man_shrugging: 
I remember that there was some discussion between @rajivnathan and @alexeykazakov about using map, and they decided for using the array. I like the array more.